### PR TITLE
Implement parsing and semantic analysis for associated consts

### PR DESCRIPTION
**Pull request by tritao**
## Description

This splits off the parsing and semantic analysis changes out of https://github.com/FuelLabs/sway/pull/4036 into a bunch of commits:

[Implement associated constants in traits and ABIs.](https://github.com/FuelLabs/sway/commit/c333b4d7409af6e278b07e09589d825e0277f5d2)

[Added multiple definition error name checking to traits.](https://github.com/FuelLabs/sway/commit/88dc9d14306ade99d0c0691e9d63a2dddb8fc6aa)

[Refactor namespace method resolving code to be item-based.](https://github.com/FuelLabs/sway/commit/5d2eb6f1770043c4218ea07a5186a8b642d79c56)

[Add constants to the trait namespace and map to allow referencing from methods](https://github.com/FuelLabs/sway/commit/d4c48f72103d8e1904e3e12fafe6f5d3059cee52) 

This one adds a bunch of disabled tests, next PR will enable them with corresponding IR generation changes and docs.

Related to https://github.com/FuelLabs/sway/issues/3797.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by otrho — DISMISSED** [review]
OK, happy to approve, although I'm not a front-end expert any more, but if this is just stage one to get parsing working and the parser is working with the tests (even though they're actually disabled) then utACK.

File: sway-core/src/language/ty/declaration/constant.rs [hunk 1]
```diff
@@ -16,37 +17,47 @@ pub struct TyConstantDeclaration {
...
     fn eq(&self, other: &Self, engines: Engines<'_>) -> bool {
+        let type_engine = engines.te();
         self.call_path == other.call_path
             && self.value.eq(&other.value, engines)
             && self.visibility == other.visibility
             && self.type_ascription.eq(&other.type_ascription, engines)
             && self.is_configurable == other.is_configurable
+            && type_engine
+                .get(self.return_type)
+                .eq(&type_engine.get(other.return_type), engines)
```
Comment (mohammadfawaz) [inline sway-core/src/language/ty/declaration/constant.rs:37]: Is `self.implementing_type` not relevant here?

File: sway-core/src/language/ty/declaration/constant.rs [hunk 2]
```diff
@@ -16,37 +17,47 @@ pub struct TyConstantDeclaration {
...
             // these fields are not hashed because they aren't relevant/a
             // reliable source of obj v. obj distinction
             attributes: _,
             span: _,
+            implementing_type: _,
         } = self;
         call_path.hash(state);
         value.hash(state, engines);
         visibility.hash(state);
+        type_engine.get(*return_type).hash(state, engines);
```
Comment (mohammadfawaz) [inline sway-core/src/language/ty/declaration/constant.rs:60]: Same question here.. What about `implementing_type`? Can't we have two identical constants but implemented differently for two different types?

File: sway-core/src/language/ty/declaration/trait.rs [hunk 3]
```diff
@@ -128,9 +142,15 @@ impl SubstTypes for TyTraitDeclaration {
...
                     let new_decl_ref = function_decl_ref
                         .clone()
-                        .subst_types_and_insert_new(type_mapping, engines);
+                        .subst_types_and_insert_new_with_parent(type_mapping, engines);
                     function_decl_ref.replace_id((&new_decl_ref).into());
                 }
+                TyTraitInterfaceItem::Constant(decl_ref) => {
+                    let new_decl_ref = decl_ref
+                        .clone()
+                        .subst_types_and_insert_new(type_mapping, engines);
```
Comment (mohammadfawaz) [inline sway-core/src/language/ty/declaration/trait.rs:151]: Should this be `subst_types_and_insert_new_with_parent` like above? I'm not sure what the difference is.

File: sway-core/src/language/ty/declaration/trait.rs [hunk 4]
```diff
@@ -155,9 +176,15 @@ impl ReplaceSelfType for TyTraitDeclaration {
...
                     let new_decl_ref = function_decl_ref
                         .clone()
-                        .replace_self_type_and_insert_new(engines, self_type);
+                        .replace_self_type_and_insert_new_with_parent(engines, self_type);
                     function_decl_ref.replace_id((&new_decl_ref).into());
                 }
+                TyTraitInterfaceItem::Constant(decl_ref) => {
+                    let new_decl_id = decl_ref
+                        .clone()
+                        .replace_self_type_and_insert_new(engines, self_type);
```
Comment (mohammadfawaz) [inline sway-core/src/language/ty/declaration/trait.rs:185]: Same question. Should this be `replace_self_type_and_insert_new_with_parent`?

File: sway-core/src/decl_engine/functional_decl_id.rs [hunk 5]
```diff
@@ -10,6 +10,7 @@ use crate::{
 pub enum FunctionalDeclId {
     TraitFn(DeclId<ty::TyTraitFn>),
     Function(DeclId<ty::TyFunctionDeclaration>),
+    Constant(DeclId<ty::TyConstantDeclaration>),
```
Comment (mohammadfawaz) [inline sway-core/src/decl_engine/functional_decl_id.rs:13]: As discussed before, the name of this `enum` is now a bit confusing. You may want to create an issue to decide on a new name.

File: sway-core/src/decl_engine/ref.rs [hunk 6]
```diff
@@ -298,13 +331,17 @@ impl ReplaceDecls for DeclRefFunction {
     fn replace_decls_inner(&mut self, decl_mapping: &DeclMapping, engines: Engines<'_>) {
         let decl_engine = engines.de();
         if let Some(new_decl_ref) = decl_mapping.find_match(self.id.into()) {
-            self.id = new_decl_ref;
+            if let FunctionalDeclId::Function(new_decl_ref) = new_decl_ref {
```
Comment (mohammadfawaz) [inline sway-core/src/decl_engine/ref.rs:334]: Do we have to also consider `TraitFn` here or not really?

File: sway-core/src/decl_engine/ref.rs [hunk 7]
```diff
@@ -298,13 +331,17 @@ impl ReplaceDecls for DeclRefFunction {
...
+                self.id = new_decl_ref;
+            }
             return;
         }
         let all_parents = decl_engine.find_all_parents(engines, &self.id);
         for parent in all_parents.iter() {
             if let Some(new_decl_ref) = decl_mapping.find_match(parent.clone()) {
-                self.id = new_decl_ref;
+                if let FunctionalDeclId::Function(new_decl_ref) = new_decl_ref {
```
Comment (mohammadfawaz) [inline sway-core/src/decl_engine/ref.rs:342]: Same question: do we have to also consider `TraitFn` here or not really?

File: sway-core/src/semantic_analysis/ast_node/declaration/abi.rs [hunk 8]
```diff
@@ -73,6 +73,10 @@ impl ty::TyAbiDeclaration {
                         ctx.decl_engine.insert(method),
                     ));
                 }
+                TraitItem::Constant(_) => {
+                    // not implemented for now since abis and traits will be unified soon
```
Comment (mohammadfawaz) [inline sway-core/src/semantic_analysis/ast_node/declaration/abi.rs:77]: I think @anton-trunov said we may not be doing that for a while. Is it difficult to fill this in in a follow up PR?

**Review by mohammadfawaz — COMMENTED** [review]
Just a few comments/questions :)

**Comment by mohammadfawaz** [comment]
Oh you have merge conflicts now as well :(

**Review by esdrubal — APPROVED** [review]
LGTM.

**Review by mohammadfawaz — APPROVED** [review]

**Comment by mitchmindtree** [comment]
Nice one @tritao, excited for this!

**Pull request merged**